### PR TITLE
1st pass at coordinated failover

### PR DESCRIPTION
**Pull request by ryanlecompte**
1st pass at coordinated failover. The basic idea is:

Client-side:
- Each client joins the 'client presence' ZK group.
- When a client gets a znode deletion for the redis node config, it joins the 'client failover ack group'.
- When the client subsequently gets a znode creation callback, it leaves the 'client failover ack group'.

Node manager side:
- When the NM wants to promote a new master, it grabs references to the ZK 'client presence' and 'client failover ack' groups.
- It figures out how many clients are currently in the 'client presence' group.
- It then deletes the znode.
- It then waits for the clients to all move to the 'client failover ack' group.
- Once all of the clients have moved to the ack group or the timeout is reached, the failover is performed.

A new configuration can be set that indicates the timeout for waiting for all clients to move over to the ACK group.

### CODE REVIEW

File: lib/redis_failover/client.rb [hunk 1]
```diff
@@ -157,13 +158,59 @@ def setup_zk
...
       @zk.stat(@znode, :watch => true)
+      join_presence_group
       update_znode_timestamp
     end
 
+    # Joins the ZK presence group.
+    def join_presence_group
+      @presence_group_lock.synchronize do
+        return if @presence_group
+        @presence_group = ZK::Group.new(@zk, CLIENT_PRESENCE_GROUP)
```
Comment (eric) [inline lib/redis_failover/client.rb:181]: The group name should be based on the name of the redis_failover root node, right?

File: lib/redis_failover/client.rb [hunk 2]
```diff
@@ -157,13 +158,59 @@ def setup_zk
...
+        logger.info("Successfully joined ZK group #{CLIENT_PRESENCE_GROUP}")
+      end
+    rescue => ex
+      logger.error("Failed to join ZK group #{CLIENT_PRESENCE_GROUP}: #{ex.inspect}")
+      logger.error(ex.backtrace.join("\n"))
+    end
+
+    # Joins the ZK failover ACK group.
+    def join_failover_ack_group
+      failover_ack_group = ZK::Group.new(@zk, CLIENT_FAILOVER_ACK_GROUP)
```
Comment (eric) [inline lib/redis_failover/client.rb:193]: Same with this one — it should be based on the redis_failover root node, right?

Comment (ryanlecompte) [inline lib/redis_failover/client.rb:193]: Good points. Right now there is no notion of a redis_failover root node.
The configuration option that redis_failover currently supports is simply
the top-level znode for storing the redis nodes config. Nothing gets nested
under that, currently. However, by default the zk-groups gem will store the
group name under "/_zk/groups". So we will have:

/_zk/groups/redis_failover_client_presence
/_zk/groups/redis_failover_client_failover_ack

In the separate distributed_monitoring branch I've introduced a root
redis_failover path under which everything gets nested/based. At some point
in the future I'll nest it all under that.

**Comment by eric** [comment]
I think it would be worth taking the synchronizing code and putting it into its own class.

Here's a [description of a barrier implementation](http://zookeeper.apache.org/doc/r3.2.2/zookeeperTutorial.html#sc_barriers) from the zookeeper docs.

Instead of specifying a `size` of the barrier, it would make sense to use the current size of a `ZK::Group` to indicate the number of items in the barrier and then use another `ZK::Group` to actually implement the barrier.

I'm still trying to come up with the exact API for it, but it seems like it should have methods including `notify()`, `enter()`, and `leave()`.

**Comment by ryanlecompte** [comment]
I had the same thought about breaking out the synchronization code into its own class. I'm not a big fan of long methods, and this one got a bit too long for my tastes. I think it's a nice generalization/encapsulation to have a barrier that's implemented with the ZK group. I'll do that.

**Comment by ryanlecompte** [comment]
Hmm, I'm wondering if this is more of a "latch" than a barrier, essentially following the API here:

https://github.com/tenderlove/tusk/blob/master/lib/tusk/latch.rb

Basically you would create a Latch with the size being the current number of clients in the presence group. Then, the node manager would call #await which would simply wait (with a timeout) until all the clients have joined the ACK group (which would happen by calling #release from the client side).

Once all clients have joined the ACK group or the deadline/timeout is reached, the #await method would return and the manager could continue with the failover.

Thoughts?

**Comment by ryanlecompte** [comment]
I just refactored the node manager's group synchronization logic into a new Latch class that's backed by ZK groups. The commit is here: af870ed06422e165fdf7d271a3724c108f670bc6

Hopefully this is a bit cleaner and more reusable than what we had before.

**Comment by ryanlecompte** [comment]
Closing this pull request. Coordinated failover will be part of a future redis_failover release, and will take advantage of new recipes in the ZK gem (barrier support, etc).

**Pull request closed without merging**